### ISSUE TITLE
Modern GUI (window manager, toolkit, etc)

### ISSUE THREAD

**Issue by ids1024** [redox-os/redox#696]
#569 discusses ways to produce a modern terminal interface (which is a great idea, by the way), but I haven't seen any discussion of a modern graphical interface.

The current interface provided by orbital is usable, but primitive. It is just a typical Windows 95-like interface with a start menu equivalent at the bottom left, windows stacked on top of each other with an exit button at the top right, etc. This is fine for now, but ideally Redox would offer something better.

This issue has received more mainstream attention than a modern terminal, and software on other operating systems has done various things to modernize their user interfaces. Gnome 3, Unity, and Windows 8/10 stand out as examples.

Of course, there is not really any consensus on what the best interface would be, and the above mentioned examples all received a fair amount of criticism. I personally prefer tilling window managers over stacking window managers, but others will likely disagree.

One feature that seems quite nice to me is (some varient of) Unity's [HUD](https://wiki.ubuntu.com/Unity/HUD), but frankly I haven't used Unity enough to see if this is really so useful. I certainly think the general idea of replacing the old menu bar paradigm with something newer is a good idea at least.

I'm not suggesting that this should be a huge priority, but it is worth discussing anyway.

**Comment by artsi0m** [comment]
Hud like in unity, but in i3 and xfce.
https://github.com/jamcnaughton/i3-hud-menu/blob/master/README.md
 Also I see one cool thing: in plan9 os you can open graphical window right in terminal window.
Sorry Im not redox-developer but I think it will be intersting os.

**Comment by rushsteve1** [comment]
I think that the best part of Unity's UI was that dashboard. It was so simple and easy yet provided a necessary function that both regular and power users could find use in.

I think we should also incorporate something like OSX's spotlight or KDE's Krunner. Specifically one of Krunner's newer features where you can use to to make documents or start emails or other common actions from within it.

I think what really needs an update is the Applications menu. While the start menu style is tried and true there is a lot we could improve on.

Maybe we should start a Orbital UI features Wishlist just to list off ideas.

**Comment by 1011X** [comment]
I think rather than sticking to only a tiled window manager, we should do something similar to what `dwm` does and allow different kinds of layouts, since the best layout really depends on the application being used (e.g. GIMP, modal or dialog windows, etc., would all be better in a floating window layout rather than a tiled one).

Something like Unity's feature of searching application menus seems good. Maybe there could also be a visual indicator that such a feature exists, like a search icon/field next to the menus, so it's easier to discover.

I really like Unity's dashboard too, though the information could be presented better. The Krunner feature that @rushsteve1 mentioned sounds like the special actions that show up when you right-click an app on Unity's app bar, which I never use because I always forget they exist. :stuck_out_tongue: So instead of hiding them in a context menu, we could include those actions in the dashboard so people at least see them.

Also, there's something I've been thinking about: perhaps we could use this opportunity to replace the desktop metaphor with something different? And better? Because I might have an idea for how to go about this!

**Comment by rushsteve1** [comment]
@1011X If you have a better idea let's hear it! It's already an "experimental" OS, its the perfect place for new ideas.

**Comment by 1011X** [comment]
Awesome! OK, so first the reason I want to rethink the [desktop metaphor](https://en.wikipedia.org/wiki/Desktop_metaphor) is because it's too office-themed, which are usually bland and boring, where you're usually doing work you don't want to do.

An idea I've had was a "room metaphor", where a user's "room" would be the desktop equivalent of their home folder. 

Each user on a system would have their own room, where they can keep all their stuff. Instead of calling it a desktop, we say it's a wall, where you can set a *wallpaper*. Maybe even have widgets for stuff like calendars or sticky notes to be placed on it.

Opening the dashboard would allow you to immediately see the important and frequently-used items in your room, and even search for any item. A "file" is just an item. "Folders" could be boxes instead, where we group items together. Logging in or out could be considered entering or exiting the room. Virtual machines would just be... a computer in your room!

I kinda like this metaphor because it's general as to what its purpose is, and fits in well with how users already use computers.

**Comment by kameko** [comment]
isn't that just renaming all the concepts of the desktop metaphor? or do you perhaps want to resurrect Microsoft Bob?
Being different for the sake of being different never works out. New ideas should be practical and spawned from an issue with existing ideas, not forced into existence on a whim. It just annoys people.
just my two cents.

**Comment by ticki** [comment]
@1011X Sorry, but renaming adds nothing but confusion.

**Comment by ca1ek** [comment]
> It is just a typical Windows 95-like interface

I agree with @kameko 
Instead of listing the great new ideas for a revolutionary desktop, let's first list what is missing from the current interface, to get it on par with W95. Then, list the wrong things with W95 and fix these.

I'll start: 
- where's my taskbar
- where are my resizable windows
- where's my minimize button
- where's my system tray

@1011X 
That's a nice metaphor, but, what new does it add? It's just relabeling existing concepts.

**Comment by Nogbit** [comment]
Instead of getting it on par with W95 why not apply some strategic thinking to this otherwise we end up with something that has a little bit of everything that no body likes and looks "me too".  Granted, in these early days that may be asking too much.

Things to consider...

- Devices...so many of them and constantly connecting and disconnecting (cell phones, usb, bluetooh and more)
- Internet/Cloud....kind of like devices though not limited to connected hardware
- Services (SaaS)...music, photos, video, current desktop OS's seem too limited here

ChromeOS I think may have the right idea in today's connected world, though hardware interop sucks with ChromeOS outside of the hardware it ships on.

I think the approach would have to evolve overtime as the OS and it's features grow, but I would encourage thinking beyond a traditional desktop OS window manager.

This comes from someone who loves i3 and uses it at work.

**Comment by dumindu** [comment]
I also had the same question since the beginning. why next generation OS follows the designs of the legacy OSs, on the era of "Design is everything"? Because when you settle down on current UI elements, there will not be a go back.  
I am also an i3 and bspwm lover and think next generation UIs should inspired from [Hub Ui Concepts](https://www.pinterest.com/search/pins/?q=hud%20ui&rs=typed&term_meta[]=hud%7Ctyped&term_meta[]=ui%7Ctyped) . If you can choose this type of UI, definitely this will be the next generation OS.

**Comment by juxtaposition-rocks** [comment]
The Redox "Book" says "Generally, we stick to well-tested and proven correct designs. If it ain't broken don't fix it." so i wonder if any non-well-proven UI paradigms have much of a chance... though personally i would also like to see some experimentation in this area.

The desktop metaphor was introduced in a time when the kinds of data and the quantities of data stored on a PC were similar to the kinds and quantities of data you would find in a pre-PC-era  desk. The desktop metaphor simply does not scale well beyond this, and operating systems have been moving away from presenting a "document-centric" view of the world (documents are the focus & apps are there to work with them) to providing an "app-centric" view (you don't work with documents/data outside the context of an app). Mobile made this transition in one big reboot (at the UI level anyway). The desktop is evolving more slowly, and none of the newer designs seem "proven" yet for that form factor. The question is: is the current state of the art broken enough for Redox to come in and fix it? :-)

I, for one, would not be troubled by the thought of a Redox that doesn't have a "file system"--even under the hood. We already have "Everything is a URL" instead of "Everything is a file". In mobile, the file system is there but hidden. In Redox perhaps it could be gone completely. On the other hand, Redox is probably already too heavily invested in the file system mentality for at least two reasons. (1) Because such a change would not only affect the GUI, but also affect the "well-tested and proven" command line at the Unix-like core. And (2) because Redox is built using Rust, which uses code stored in files rather than in a database (another barbaric tradition from the age of the crufty old desktop metaphor).

**Comment by ids1024** [comment]
@juxtaposition-rocks You have a good point about the app metaphor replacing the traditional desktop metaphor on certain form factors, but I don't really see that it is any better. It makes it harder to share data between programs, and is ultimately just obscuring the hierarchical file system, which such systems are still ultimately based on. The app metaphor also generally seems more oriented towards less technical users.

**Comment by oboff** [comment]
Windows 95 desktop isn't broken and doesn't need fixing.

**Comment by learneman** [comment]
Would be cool to see something like budgie desktop on redox.  Only thing I would add to budgie
desktop is the ability to search for files in the menu.

**Comment by yohanesu75** [comment]
I would like to have a status bar in console-mode showing time, battery status, current vt# etc. A lot of work is done in console and often on laptop so this would be very handy :)

**Comment by yohanesu75** [comment]
To make everyone happy, design a decent Wayland-like compositor protocol and let users create their own desktop environment / compositors (but keep it tight to avoid fragmentation like Gnome/KDE/etc).

**Comment by torkleyy** [comment]
I think it would be great to have some text box appearing at the top of the screen and you just type something to launch an application. Probably the "Everything is an URL" concept could be integrated into this, however I'm not sure how it would work out exactly because most users won't be happy to always give a complete URL.

**Comment by darkelement** [comment]
Hi! I'm developer of `perceptia` (https://github.com/perceptia/perceptia) - dynamic window manager with support for Wayland. `perceptia` is focused on implementing all functionality needed to create display server in Rust. I'd like to explore idea of porting it to Redox.

I could not find any specifics how authors of Redox imagine its GUI future. If that aligns with their vision I could work on support for Wayland in Redox. Porting window manager would be good start. Later maybe providing Wayland or Wayland-like support in orbtk - depending on goals and priorities of Redox.

Let me know if anyone is interested.

**Comment by Pharaone** [comment]
@darkelement good project!

**Comment by spease** [comment]
I might be interested in helping with Redox UI stuff too. :)

**Comment by minecrawler** [comment]
Imho, I really love the idea behind Redox. Having a very stable system, which uses proven, intuitive concepts users can directly recognize and love, is a very good base. See Debian, which is also very conservative. So many other systems build on top of it because of that! That's the Redox I was introduced and that's the Redox I want to see. It would allow the average user to switch from their Windows to Redox easily in terms of usability.

At the same time, my desktop PC is rolling KDE with lots of "unstable" stuff. I love to tinker. I am not your average user. I want more. I need a KDE DE, which allows me to completely customize the whole desktop experience. I can remove or display stuff on the desktop. I can decide whether I want a start button. How many taskbars and where. Which icons I find important. Anything imaginable is possible with KDE. That's not the proven way, that's way over top and confusing for normal users and not something I see in scope for Redox OS.

Back in 2016, at FOSDEM, I asked the Redox Team, if it was possible to use the Redox kernel in the same way the Linux kernel is used: To create new distributions. Their answer was: Yes, go for it! Now, some of you might disagree, however, in my opinion, Redox OS made its way clear. They want a proven system. If that leads to a Win95 kind of system, then that's ok.

For anyone else, who, like me, wants to have something else, there is always the possibility to create a new distribution. Call it fragmentation, but it's one of the things I love about the Linux world. My stationary is my tinker-baby (Gentoo), my laptop is on the  edge, but rock-solid (KaOS), my Pi stable and only needs a few MB for its system (StaLi) and my server is conservative and just works (Debian/NixOS). Sure, there are flaws, but there will always be ideas to improve existing things. Let's get something awesome started here! This is Redox OS, so let's do it the Redox way (so no perceptia either). Let's keep our wishes for the custom distributions we will be able to create and let's have live our wet dreams live there!

**Comment by nixpulvis** [comment]
I see this issue simply as a desire to see another crate for a "better" Orbital implementation for users who like tiling window managers like `i3`. I feel like the core Redox project shouldn't make any claims about Graphical User Interface.

**Comment by InsidiousMind** [comment]
I'd like to add it may be worth choosing a subset of features, then making these very polished (and visually stimulating, if Redox goes for that) being customizable and *functional*. IE a hud UI 'concept' might look really cool but not be functional whatsoever.

A DE/WM with feature bloat might seem cool on paper but might be annoying to use. In that respect, I'd opt for something closer to i3 than KDE.

I would go with possibly making a nice hub-style layout that is hotkey and mouse-friendly but includes such things as fuzzy-searching for files, applications etc while being tightly integrated with Redox's planned/existing core group of applications (Photo, Video, Music, whatever). I think Redox is great, and maybe our thinking should be 'we now have the tools to innovate on today's OS design, so how will we approach this?' rather than 'lets tack on even more stuff to be cool'. EX: what kind of design features can be created from 'Everything is a URL' rather than 'Everything is a file'. In virtually every OS today a file is represented like those of the filing cabinets of yore. Will we choose to represent them in the same way?

However, I'm not sure we are there yet. I agree with @ca1ek. It's definitely worth polishing off the features that are missing. Better to have something functional then over-bloated and broken.

**Comment by cookie545445** [comment]
There is a lot of needless reinvention here. I get that Redox is supposed to be new and have some sort of uniqueness and advantage over everything else, but that doesn’t mean we can’t take things from elsewhere.

In this thread I've seen:
* renaming of already existent concepts and their components, and suggestion of these renamed things over the things with the old names
* "we should make a decent Wayland-like protocol" I've heard of one of those, it's called Wayland
* "we should get rid of the filesystem" how would this help to achieve anything?

Redox already has advantages over other OSes without even considering anything outside kernel space: namespaces, the microkernel design, the usage of a safe language. It's not like we need new ideas that are as ridiculous as some of the suggestions here are to stand out. Standing out isn't even always a good thing.

TL;DR: Please don't suggest new things just for the sake of novelty.

**Comment by that1communist** [comment]
@InsidiousMind 

What about i3's window management with KDE's panel?

Just enabling floating by default would make it user friendly enough, while still as powerful as i3 and kde.

**Comment by nixpulvis** [comment]
I think as a matter of priciple we can name things whatever we want, but there are places where adhering to an existing protocol might yield both practical and "theoretical" simplicity, for example avoiding tieing Redox too tightly to contentious topics like UI and window managers.

**Comment by BadFactor** [comment]
First of all, Linux OS System is great, but we should keep an Ey on it, Otherwise we have dozen of Distributions that neuer Seen Desktop PC in life Time, like Linux! We should lern from failures and not make the Same! The mistakes that the Linux World makes good is that it offers us "Open" Software and Modularity, Rock solid System. The World is using Windows as there Desktop System, because Its easy to use! On the otherhand Its not easy for a noobie to go pro!  Lots of hidden options aso. Thousands of clicks before u found what ur looking for. The Yast Interface from, Open suse /suse Linux is somekind of perfect! Lets See the overhaul look, some Users dont know what they can do with there PC, some are really crazy on it, If u Design a new User Interface, u have to, Maybe Design three in one!  One for "dumb" people or they dont care really much on it, then the ongoing pro people and the pro peoples! The really next gen ui is something with a super intelligent Assistant, Maybe showed up on screen like an 80's 3d Head Character type!  That can be Controlled, with voice commands, eye contact, aso but without a mice or Keyboard! 

The only hope is that it will not going to be a second Linux and it will have build in vulkan API,  Maybe via vulkano, and some other good. Software in the Rustlang Organisation/ Group / foundation whatever u wanna Calls it

**Comment by BadFactor** [comment]
Sorry for grammar aso. I just needed to say, build a tiled and stacked Ui in one with upgradable Interfaces like the one i write before,  3d Character aso u know what i mean? Going online is good, but going offline is the Same? Dont going to be a online OS Its nonsense, If u know, i mean knowing, just offline System works for some Time, there are pretty awesome for that Time, but the Same can be the Same way with just online OS, that kills everything AT some point! Balance is the key! 

The other topic i mentioned, Linux distros, Lets say, u guys say "okay lets do a distro, or whatever, from redox" Then it Takes not long time, to have 1.000.000 of them, or higher and thats bullshit, Lets say okay we agreed to make a Maximum of 5, then Its better to overlook, better to controlling with Bugfixes aso. Better for everybody!  Make a easy to use redox, for example then a pro Version and so on, Maybe a gaming OS for a Video console, but just for PC, so u can use ur PC as a normal gaming Video console, PlayStation, xbox, dreamcast, wii, switch aso aso aso.

**Comment by fvckya** [comment]
Hi guys, I have a desire to come up with a design for a redox ...
Since I recently in programming (most recently), I can not code with you. But I can help with ideas.

The first change of the logo, of course what I'm offering is just a sketch.
But before continuing, I would like to ask you, otherwise it will be in vain ..
there will be something similar to the logo Rust, and yours.

![logo](https://user-images.githubusercontent.com/29673961/33525542-ad29a23c-d85b-11e7-8ea4-dfa3bed3debd.jpg)


The second. Own Desktop Env and tulkit.
no global menus and HUD. `Alt `button will be toogle for app menu, and window name. 
additional button next to the buttons close the window, hide/show quick buttons.

![rust](https://user-images.githubusercontent.com/29673961/33525555-f106defc-d85b-11e7-815f-634e4b4976e6.jpg)

translated with Google translation. If you like it and know anyone that knows russian, it would be great

**Comment by VoidNoire** [comment]
I agree with @nixpulvis about:
> I see this issue simply as a desire to see another crate for a "better" Orbital implementation for users who like tiling window managers like `i3`. I feel like the core Redox project shouldn't make any claims about Graphical User Interface.
and 
> avoiding tieing Redox too tightly to contentious topics like UI and window managers.

I think the RedoxOS team at most should simply go with the suggestion by @johalun instead of focusing too much on UI design: 
> To make everyone happy, design a decent Wayland-like compositor protocol and let users create their own desktop environment / compositors

TL;DR Just redesign Orbital (doesn't even have to be Wayland or Wayland-like necessarily) and write good documentation for it such that it is easier for third parties to develop UI components that can be pick and mixed by users as they see fit, as opposed to constraining them to a particular style of UI/desktop environment.
